plugin/vcs/github/github.go

```diff
@@ -0,0 +1,286 @@
+	return errors.New("not implemented yet") // TODO: https://github.com/bytebase/bytebase/issues/928
+}
+
+func (p *Provider) DeleteWebhook(ctx context.Context, oauthCtx common.OauthContext, instanceURL, repositoryID, webhookID string) error {
+	return errors.New("not implemented yet") // TODO: https://github.com/bytebase/bytebase/issues/928
+}
+
+func httpGet(ctx context.Context, client *http.Client, resourcePath string, token *string, oauthCtx oauthContext, refresher common.TokenRefresher) (code int, respBody string, err error) {
+	return retry(ctx, client, token, oauthCtx, refresher,
+		func() (*http.Response, error) {
+			url := fmt.Sprintf("%s/%s", apiURL, resourcePath)
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+			if err != nil {
+				return nil, errors.Wrapf(err, "construct GET %s", url)
+			}
+
+			req.Header.Set("Content-Type", "application/json")
+			req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", *token))
+			resp, err := client.Do(req)
+			if err != nil {
+				return nil, errors.Wrapf(err, "GET %s", url)
+			}
+			return resp, nil
+		},
+	)
+}
+
+const maxRetries = 3
+
+func retry(ctx context.Context, client *http.Client, token *string, oauthCtx oauthContext, refresher common.TokenRefresher, f func() (*http.Response, error)) (code int, respBody string, err error) {
+	var resp *http.Response
+	var body []byte
+	for retries := 0; retries < maxRetries; retries++ {
+		select {
+		case <-ctx.Done():
+			return 0, "", ctx.Err()
+		default:
+		}
+
+		resp, err = f()
+		if err != nil {
+			return 0, "", err
+		}
+
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return 0, "", errors.Wrapf(err, "read response body with status code %d", resp.StatusCode)
+		}
+
+		if err = getOAuthErrorDetails(resp.StatusCode, string(body)); err != nil {
+			if _, ok := err.(*oauthError); ok {
+				// Refresh and store the token.
+				if err := refreshToken(ctx, client, token, oauthCtx, refresher); err != nil {
+					return 0, "", err
+				}
+				continue
+			}
+			return 0, "", errors.Errorf("want oauthError but got %T", err)
+		}
+		return resp.StatusCode, string(body), nil
+	}
+	return 0, "", errors.Errorf("retries exceeded for oauth refresher with status code %d and body %q", resp.StatusCode, string(body))
+}
+
+type oauthError struct {
+	Err              string `json:"error"`
+	ErrorDescription string `json:"error_description"`
+}
+
+func (e oauthError) Error() string {
+	return fmt.Sprintf("GitHub oauth response error %q description %q", e.Err, e.ErrorDescription)
+}
+
+// getOAuthErrorDetails only returns error if it's an OAuth error. For other
+// errors like 404 we don't return error. We do this because this method is only
+// intended to be used by oauth to refresh access token on expiration.
+func getOAuthErrorDetails(code int, body string) error {
+	if code < http.StatusMultipleChoices {
```

Comment:
The `if` condition in GitLab implementation looks weird:

https://github.com/bytebase/bytebase/blob/83f3f4dcca7d7c65929af3e8b13dd059c168c66e/plugin/vcs/gitlab/gitlab.go#L728

If `code < 300` is true, `200 <= code` is always true.

Reply:
code = 100
code < 300 -> true
200 <= code -> false

Reply:
You're right 🤦 I must be drunk.